File: engine/src/analyze/importResolver.ts
export interface ImportResolverConfig {
  sourceFiles: Set<string>;
  tsconfigPaths?: Record<string, string[]>;
}

function normalizePathComponents(path: string): string {
  return path.split("/").filter((component) => component.length > 0).join("/");
}

export class ImportResolver {
  private readonly sourceFiles: Set<string>;
  private readonly tsconfigPaths: Record<string, string[]>;
  private readonly extensions = ["", ".ts", ".tsx", ".js", ".jsx", ".mjs", ".py", ".swift", ".go"];

  constructor(config: ImportResolverConfig) {
    this.sourceFiles = new Set(
      [...config.sourceFiles].map((path) => normalizePathComponents(path)),
    );
    this.tsconfigPaths = config.tsconfigPaths ?? {};
  }

  resolve(specifier: string, sourcePath: string): string | null {
    if (specifier.startsWith(".")) {
      return this.resolveRelative(specifier, sourcePath);
    }

    const aliased = this.resolveAlias(specifier);
    if (aliased !== null) {
      return aliased;
    }

    return this.resolveCandidate(specifier);
  }

  private resolveRelative(specifier: string, sourcePath: string): string | null {
    const sourceDirectory = deletingLastPathComponent(sourcePath);
    return this.resolveCandidate(this.normalize(`${sourceDirectory}/${specifier}`));
  }

  private resolveAlias(specifier: string): string | null {
    for (const [pattern, targets] of Object.entries(this.tsconfigPaths)) {
      const captured = this.captureAliasWildcard(specifier, pattern);
      if (captured === null) {
        continue;
      }

      for (const target of targets) {
        const candidate = target.replace(/\*/g, captured);
        const resolved = this.resolveCandidate(candidate);
        if (resolved !== null) {
          return resolved;
        }
      }
    }

    return null;
  }

  private resolveCandidate(candidate: string): string | null {
    const normalizedCandidate = this.normalize(candidate);

    for (const suffix of this.extensions) {
      const path = normalizedCandidate + suffix;
      if (this.sourceFiles.has(path)) {
        return path;
      }
    }

    for (const suffix of this.extensions.slice(1)) {
      const indexPath = `${normalizedCandidate}/index${suffix}`;
      if (this.sourceFiles.has(indexPath)) {
        return indexPath;
      }
    }

    return null;
  }

  private captureAliasWildcard(specifier: string, pattern: string): string | null {
    const wildcardIndex = pattern.indexOf("*");
    if (wildcardIndex === -1) {
      return specifier === pattern ? "" : null;
    }

    const prefix = pattern.slice(0, wildcardIndex);
    const suffix = pattern.slice(wildcardIndex + 1);

    if (!specifier.startsWith(prefix) || !specifier.endsWith(suffix)) {
      return null;
    }

    return specifier.slice(prefix.length, specifier.length - suffix.length);
  }

  private normalize(path: string): string {
    const components: string[] = [];

    for (const component of path.split("/").filter((value) => value.length > 0)) {
      switch (component) {
        case ".":
          continue;
        case "..":
          components.pop();
          break;
        default:
          components.push(component);
      }
    }

    return components.join("/");
  }
}

/** Mirrors `NSString.deletingLastPathComponent` for the relative paths used here. */
function deletingLastPathComponent(path: string): string {
  const index = path.lastIndexOf("/");
  return index >= 0 ? path.slice(0, index) : "";
}

File: app/Sources/GunkApp/Analyze/ImportResolver.swift
import Foundation

struct ImportResolver: Sendable {
  struct Config: Equatable, Sendable {
    let sourceFiles: Set<String>
    let tsconfigPaths: [String: [String]]

    init(sourceFiles: Set<String>, tsconfigPaths: [String: [String]] = [:]) {
      self.sourceFiles = Set(sourceFiles.map(Self.normalize(path:)))
      self.tsconfigPaths = tsconfigPaths
    }

    private static func normalize(path: String) -> String {
      path.split(separator: "/", omittingEmptySubsequences: true).joined(separator: "/")
    }
  }

  private let config: Config
  private let extensions = ["", ".ts", ".tsx", ".js", ".jsx", ".mjs", ".py", ".swift", ".go"]

  init(config: Config) {
    self.config = config
  }

  func resolve(_ specifier: String, from sourcePath: String) -> String? {
    if specifier.hasPrefix(".") {
      return resolveRelative(specifier, from: sourcePath)
    }

    if let aliased = resolveAlias(specifier) {
      return aliased
    }

    return resolveCandidate(specifier)
  }

  private func resolveRelative(_ specifier: String, from sourcePath: String) -> String? {
    let sourceDirectory = (sourcePath as NSString).deletingLastPathComponent
    return resolveCandidate(normalize(path: "\(sourceDirectory)/\(specifier)"))
  }

  private func resolveAlias(_ specifier: String) -> String? {
    for (pattern, targets) in config.tsconfigPaths {
      guard let captured = captureAliasWildcard(specifier: specifier, pattern: pattern) else {
        continue
      }

      for target in targets {
        let candidate = target.replacingOccurrences(of: "*", with: captured)
        if let resolved = resolveCandidate(candidate) {
          return resolved
        }
      }
    }

    return nil
  }

  private func resolveCandidate(_ candidate: String) -> String? {
    let normalizedCandidate = normalize(path: candidate)

    for suffix in extensions {
      let path = normalizedCandidate + suffix
      if config.sourceFiles.contains(path) {
        return path
      }
    }

    for suffix in extensions.dropFirst() {
      let indexPath = "\(normalizedCandidate)/index\(suffix)"
      if config.sourceFiles.contains(indexPath) {
        return indexPath
      }
    }

    return nil
  }

  private func captureAliasWildcard(specifier: String, pattern: String) -> String? {
    guard let wildcardRange = pattern.range(of: "*") else {
      return specifier == pattern ? "" : nil
    }

    let prefix = String(pattern[..<wildcardRange.lowerBound])
    let suffix = String(pattern[wildcardRange.upperBound...])

    guard specifier.hasPrefix(prefix), specifier.hasSuffix(suffix) else {
      return nil
    }

    return String(specifier.dropFirst(prefix.count).dropLast(suffix.count))
  }

  private func normalize(path: String) -> String {
    var components: [String] = []

    for component in path.split(separator: "/", omittingEmptySubsequences: true).map(String.init) {
      switch component {
      case ".":
        continue
      case "..":
        _ = components.popLast()
      default:
        components.append(component)
      }
    }

    return components.joined(separator: "/")
  }
}
